src/components/context/TodosContext.tsx

```diff
@@ -0,0 +1,230 @@
+    try {
+      await todosServices.deleteTodos(todoId);
+      setTodos(currTodos => currTodos.filter(todo => todo.id !== todoId));
+    } catch (error) {
+      setErrorMessage(TodoError.UnableToDelete);
+      throw error;
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  const updateTodo = async (updatedTodo: Todo) => {
+    try {
+      const todoEdited = await todosServices.updateTodo(updatedTodo);
+
+      setTodos(currTodos =>
+        currTodos.map(todo => (todo.id === updatedTodo.id ? todoEdited : todo)),
+      );
+    } catch (error) {
+      setErrorMessage(TodoError.UnableUpdate);
+      throw error;
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  const handleCheck = async (updatedTodo: Todo) => {
+    setIsLoading(true);
+    try {
+      await updateTodo({ ...updatedTodo, completed: !updatedTodo.completed });
+    } catch (error) {
+      setErrorMessage(TodoError.UnableUpdate);
+      throw error;
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  const toggleAll = async () => {
+    const allCompleted = todos.every(todo => todo.completed);
+    const activeTodos = todos.filter(todo => !todo.completed);
+    const completedTodos = todos.filter(todo => todo.completed);
+    const todosToToggle = allCompleted ? completedTodos : activeTodos;
+    const togglePromise = todosToToggle.map(todo =>
+      todosServices.updateTodo({
+        ...todo,
+        completed: !allCompleted,
+      }),
+    );
+
+    setIsLoading(true);
+
+    try {
+      setIsLoading(true);
+
+      await Promise.all(togglePromise);
+      setTodos(
+        todos.map(todo => ({
+          ...todo,
+          completed: !allCompleted,
+        })),
+      );
+    } catch (error) {
+      setErrorMessage(TodoError.UnableUpdate);
+      throw error;
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  const handleClearCompleted = async () => {
+    const completedTodos = todos.filter(todo => todo.completed);
+    const failedDeletions: Todo[] = [];
+
+    for (const todo of completedTodos) {
+      try {
+        await todosServices.deleteTodos(todo.id);
+        setIsLoading(false);
+      } catch (error) {
+        setTodos(todos);
+        failedDeletions.push(todo);
+        setErrorMessage(TodoError.UnableToDelete);
+      }
+    }
+
+    if (failedDeletions.length === 0) {
+      setTodos(todos.filter(todo => !todo.completed));
+    } else {
+      setTodos(
+        todos.filter(
+          todo =>
+            !completedTodos.includes(todo) || failedDeletions.includes(todo),
+        ),
+      );
+      setErrorMessage(TodoError.UnableToDelete);
+    }
+  };
+
+  const contextValueMemo = useMemo(
+    () => ({
+      todos,
+      statusTodo,
+      tempTodo,
+      isLoading,
+      setIsLoading,
+      setTempTodo,
+      removeTodo,
+      addTodo,
+      updateTodo,
+      handleCheck,
+      handleClearCompleted,
+      toggleAll,
+      setStatusTodo,
+    }),
+    [todos, statusTodo, tempTodo],
```

Comment:
useMemo here does not provide a significant performance improvement in this context. But you can use useCallback for functions

Reply:
> useMemo here does not provide a significant performance improvement in this context. But you can use useCallback for functions

Good point